receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+// the instance.
+type timeseriesMapPdata struct {
+	sync.RWMutex
+	mark   bool
+	tsiMap map[string]*timeseriesinfoPdata
+}
+
+// Get the timeseriesinfo for the timeseries associated with the metric and label values.
+func (tsm *timeseriesMapPdata) get(metric *pdata.Metric, kv pdata.AttributeMap) *timeseriesinfoPdata {
+	name := metric.Name()
+	sig := getTimeseriesSignaturePdata(name, kv)
+	if metric.DataType() == pdata.MetricDataTypeHistogram {
+		// There are 2 types of Histograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		aggTemporality := metric.Histogram().AggregationTemporality()
+		sig += "," + aggTemporality.String()
+	}
+	tsi, ok := tsm.tsiMap[sig]
+	if !ok {
+		tsi = &timeseriesinfoPdata{}
+		tsm.tsiMap[sig] = tsi
+	}
+	tsm.mark = true
+	tsi.mark = true
+	return tsi
+}
+
+// Create a unique timeseries signature consisting of the metric name and label values.
+func getTimeseriesSignaturePdata(name string, kv pdata.AttributeMap) string {
+	labelValues := make([]string, 0, kv.Len())
+	kv.Sort().Range(func(_ string, attrValue pdata.AttributeValue) bool {
+		value := attrValue.StringVal()
+		if value != "" {
+			labelValues = append(labelValues, value)
+		}
+		return true
+	})
+	return fmt.Sprintf("%s,%s", name, strings.Join(labelValues, ","))
+}
+
+// Remove timeseries that have aged out.
+func (tsm *timeseriesMapPdata) gc() {
+	tsm.Lock()
+	defer tsm.Unlock()
+	// this shouldn't happen under the current gc() strategy
+	if !tsm.mark {
+		return
+	}
+	for ts, tsi := range tsm.tsiMap {
+		if !tsi.mark {
+			delete(tsm.tsiMap, ts)
+		} else {
+			tsi.mark = false
+		}
+	}
+	tsm.mark = false
+}
+
+func newTimeseriesMapPdata() *timeseriesMapPdata {
+	return &timeseriesMapPdata{mark: true, tsiMap: map[string]*timeseriesinfoPdata{}}
+}
+
+// JobsMapPdata maps from a job instance to a map of timeseriesPdata instances for the job.
+type JobsMapPdata struct {
+	sync.RWMutex
+	gcInterval time.Duration
+	lastGC     time.Time
+	jobsMap    map[string]*timeseriesMapPdata
+}
+
+// NewJobsMap creates a new (empty) JobsMapPdata.
+func NewJobsMapPdata(gcInterval time.Duration) *JobsMapPdata {
+	return &JobsMapPdata{gcInterval: gcInterval, lastGC: time.Now(), jobsMap: make(map[string]*timeseriesMapPdata)}
+}
+
+// Remove jobs and timeseries that have aged out.
+func (jm *JobsMapPdata) gc() {
+	jm.Lock()
+	defer jm.Unlock()
+	// once the structure is locked, confirm that gc() is still necessary
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		for sig, tsm := range jm.jobsMap {
+			tsm.RLock()
+			tsmNotMarked := !tsm.mark
+			tsm.RUnlock()
+			if tsmNotMarked {
+				delete(jm.jobsMap, sig)
+			} else {
+				tsm.gc()
+			}
+		}
+		jm.lastGC = time.Now()
+	}
+}
+
+func (jm *JobsMapPdata) maybeGC() {
+	// speculatively check if gc() is necessary, recheck once the structure is locked
+	jm.RLock()
+	defer jm.RUnlock()
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		go jm.gc()
```

Comment:
If `get` can be called from multiple threads, I think this can cause gc to be called multiple times concurrently, I suspect that's not intended. Can use a comment if this case can't happen since a simple method like `get` we'd generally expect to not have any threading guarantees

Reply:
The time since the last GC is checked again inside `gc()` once a lock is obtained.  It may be triggered multiple times, but it will still only run once per interval.  That's what the comments in this method and in `gc()` are trying to convey.